### PR TITLE
[JENKINS-61841] Limit the number of exceptions stored in CompositeIOException

### PR DESCRIPTION
**Pull request by dwnusbaum**
See [JENKINS-61841](https://issues.jenkins-ci.org/browse/JENKINS-61841).

This PR truncates the list of exceptions passed to `CompositeIOException` to only contain the first 15 exceptions. 15 was my rough guess for the point where additional exceptions are probably going to contain mostly redundant information. Right now, this value is not configurable by users, but I am happy to make it configurable as a system property or change the default value if desired.

### Proposed changelog entries

* Robustness: Truncate the list suppressed exceptions returned by some operations such as recursive directory deletion. Previously, in rare cases, exceptions thrown when failing to delete large directories could consume significant amounts of memory.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

File: core/src/main/java/hudson/Util.java [hunk 1]
```diff
@@ -1708,8 +1708,24 @@ public static long daysElapsedSince(@NonNull Date date){
...
+     * depending on the value of {@link DELETION_RETRIES}.
+     * <p>
+     * The purpose of this setting is to prevent the composite exceptions
+     * thrown by failed directory deletion operations from becoming extremely
+     * large. For example, in cases where Jenkins attempts to delete a large
+     * directory that it does not have permission to delete, we do not want the
+     * composite exception to contain an exception for every single file inside
+     * of that directory.
+     */
+    private static final int DELETION_MAX_ERRORS_BEFORE_ABORT = 10;
```
Comment (dwnusbaum) [inline core/src/main/java/hudson/Util.java:1725]: I am not sure whether we need to make this configurable by system property, so I did not make it configurable, but I am happy to make it configurable if there is a reasonable case for doing so. I am also happy to increase/decrease the default value, maybe 15 or 25 would be a reasonable default instead.

File: core/src/main/java/jenkins/util/io/PathRemover.java [hunk 2]
```diff
@@ -113,6 +115,14 @@ public void forceRemoveRecursive(@NonNull Path path) throws IOException {
     public interface RetryStrategy {
         boolean shouldRetry(int retriesAttempted);
 
+        /**
+         * Whether the current deletion attempt should be aborted early based on the errors seen so far.
+         * Additional attempts may still occur according to {@link #shouldRetry} even if the current attempt is aborted.
+         */
+        default boolean shouldAbort(List<IOException> accumulatedErrors) {
```
Comment (dwnusbaum) [inline core/src/main/java/jenkins/util/io/PathRemover.java:122]: I was not sure on the best way to integrate this configuration into `PathRemover`, so for now I just added it as a method in `RetryStrategy`. I could also see a case for creating a distinct `AbortStrategy` field with just this method, or just adding a `maxErrorsBeforeAbort` field directly in `PathRemover`, but that seemed like it didn't really fit the existing design.

File: core/src/main/java/jenkins/util/io/PathRemover.java [hunk 3]
```diff
@@ -209,26 +226,30 @@ public String failureMessage(@NonNull Path fileToRemove, int retryCount) {
         }
     }
 
-    private List<IOException> tryRemoveRecursive(@NonNull Path path) {
+    private void tryRemoveRecursive(@NonNull Path path, List<IOException> accumulatedErrors) {
```
Comment (dwnusbaum) [inline core/src/main/java/jenkins/util/io/PathRemover.java:229]: I modified this method and `tryRemoveDirectoryContents` to accept the list of errors as a parameter and mutate it instead of creating new lists at every level so that `shouldAbort` can act on the number of exceptions seen at every level rather than only those seen in the current directory.

File: core/src/test/java/jenkins/util/io/PathRemoverTest.java [hunk 4]
```diff
@@ -418,6 +423,34 @@ public void testForceRemoveRecursive_ParentIsSymbolicLink() throws IOException {
         assertFalse(d1.exists());
     }
 
+    @Test
+    @Issue("JENKINS-55448")
+    public void testForceRemoveRecursive_AbortsAfterTooManyErrors() throws IOException {
+        assumeTrue(Functions.isWindows());
```
Comment (dwnusbaum) [inline core/src/test/java/jenkins/util/io/PathRemoverTest.java:429]: I'm not sure how to test this behavior on Linux without a ton of mocking, which I prefer to avoid, so it's Windows-only.

File: core/src/test/java/jenkins/util/io/PathRemoverTest.java [hunk 5]
```diff
@@ -418,6 +423,34 @@ public void testForceRemoveRecursive_ParentIsSymbolicLink() throws IOException {
         assertFalse(d1.exists());
     }
 
+    @Test
+    @Issue("JENKINS-55448")
+    public void testForceRemoveRecursive_AbortsAfterTooManyErrors() throws IOException {
```
Comment (dwnusbaum) [inline core/src/test/java/jenkins/util/io/PathRemoverTest.java:428]: I just added one basic test for now, but if anyone sees an edge case that is not covered by this test or has a specific case they want to test, let me know and I can work on adding more tests.

**Comment by daniel-beck** [comment]
Did you consider not recording every exception and just keeping that list short, vs. stopping deletion attempts?

**Comment by dwnusbaum** [comment]
> Did you consider not recording every exception and just keeping that list short, vs. stopping deletion attempts?

Yeah, I think that could be fine too. There is a minor performance cost to throwing all of the exceptions and generating their stack traces that is avoided with the current approach, but I do not feel strongly about it. One thing in favor of your approach is that without my PR, in extreme cases, the time spent attempting to delete thousands of files is kind of a poor man's dynamic addition to `Util.WAIT_BETWEEN_DELETION_RETRIES`, so this change as-is might require users to increase the wait time to compensate for that not longer happening?

**Comment by daniel-beck** [comment]
Also number of retries is limited, and if the order of files is consistent, chances are better than not that subsequent attempts won't get much further. A 99% deletion may easily become a 1% deletion because of a bad subdirectory this way.

**Comment by dwnusbaum** [comment]
> A 99% deletion may easily become a 1% deletion because of a bad subdirectory this way

We could increase the default value of the constant to make this kind of pathological case less likely, but it seems better to just keep the old behavior the same to avoid any potential issues and truncate the list of exceptions, so I'll go ahead and change to that approach.

One thing I'm not sure about is whether we should handle the truncation in `PathRemover` or in `CompositeIOException` in case any other code runs into a similar issue. The only other code instantiating `CompositeIOException` right now is `LogRotator.perform`, which doesn't seem likely to run into the extreme cases described in this PR, but maybe it's better to be careful and truncate everywhere just in case?

File: core/src/main/java/jenkins/util/io/CompositeIOException.java [hunk 6]
```diff
@@ -32,25 +32,43 @@
 import java.io.UncheckedIOException;
 import java.util.Arrays;
 import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 @Restricted(NoExternalUse.class)
 public class CompositeIOException extends IOException {
-    private final List<IOException> exceptions;
```
Comment (dwnusbaum) [inline core/src/main/java/jenkins/util/io/CompositeIOException.java:38]: I wanted to remove this because it is redundant, but I realized that introduces a serialization incompatibility, so I'll revert it. I will also add an explicit `serialVersionUID`.

**Comment by dwnusbaum** [comment]
I'm just going to close the PR until I have a chance to test it on my Windows machine.

**Comment by dwnusbaum** [comment]
I filed a rebased and modified version of this PR as https://github.com/jenkinsci/jenkins/pull/4655.
